api/STRUCTURE.md

```diff
@@ -0,0 +1,83 @@
+- run-local-rw-migrations (runs migrations only on the `workbench` schema)
+
+## DAOs
+
+Data Access Objects (DAOs) are interfaces that describe what operations the application can perform on the database. Generaly, we give DAOs the CamelCased name of the database table. All of the Workbench DAOs implement Spring [`CrudRepository`](https://docs.spring.io/spring-data/commons/docs/current/api/org/springframework/data/repository/CrudRepository.html), which means they can do simple operations like find-by-primary-key, save, delete-by-primary-key, etc.
+
+CrudRepository also allows queries to be [automatically derived](https://docs.spring.io/spring-data/jpa/docs/current/reference/html/#repositories.query-methods.details) from the name of a method defined on the interface. As such, you will see methods like `findByUserIdAndWorkspaceId` with no apparent implementation. The implementation is generated by Spring.
+
+To define a query that uses a foreign key, there are a couple of options. First, you could use `@JoinColumn`/`@[One|Many]ToOne` annotation on the corresponding DB Model, as described [below](#db-models). Second, you could use an [@Query](https://www.baeldung.com/spring-data-jpa-query) annotation to just describe the query in stringy SQL.
+
+TODO: do we have a preference of which of those to do?
+
+## DB Models
+
+DB Models represent a single row of a single table in the database, possibly with extender tables attached. We denote DB models by prefixing `Db` to the name of the database table. DB models heavily use `javax.persistence` annotations to describe where in the database table each part of the model lives.
+
+In order to automatically gather data from extender tables, we use the `@JoinColumn`, `@OneToOne`, `@OneToMany`, `@ManyToOne`, and `@MappedBy` [annotations](https://www.baeldung.com/jpa-join-column). When using these annotations, be sure to carefully consider which table should 'own' the relationship. With a `@ManyToOne` relationship, the `@ManyToOne` side should generally own the relationship. For example. if we have an extender table containing a user's addresses, `DbAddress` should have a `@ManyToOne` `DbUser` member and `DbUser` should have a `@OneToMany` `DbAddress` member, with the annotation containing a `mappedBy` attribute pointing to `DbAddress`'s `DbUser` member as the member defining the relationship.
+
+TODO: wording on the above is Bad. pls help
+
+Much of the structure of DB models can be autogenerated in IntelliJ by defining the class members and then doing Generate -> Getter and Setter and Generate -> `equals()` and `hashCode()`. This will not automatically add the required annotations - those must be added manually.
+
+## Services and ServiceImpls
+
+Services are interfaces that declare methods of business logic. ServiceImpls are classes implementing their correspondingly named Service and the business logic methods declared therein.
+
+TODO: Why do we do this this way?
```

Comment:
It's 75% philosophy, 25% requirement. Basically, a Java class can implement any number of Spring beans for one or more services, and a given Spring bean can have any number of implementations.

Declaring a `@Service` directly on the implementation class is stepping on both of those degrees of freedom. A dependent service shouldn't need to know (or have any way to detect) the implementation class chosen.

I actually got bit by this in the case of the UserService/UserServiceImpl pair. Before the latter was created, there was no (direct) way for implementation context to be reused to support "sidecar services" like the `GaugeDataCollector` bean.